### ISSUE TITLE
Pycodestyle as a library: getting configuration?

### ISSUE THREAD

**Issue by HiPhish** [PyCQA/pycodestyle#1010]
Hello,

I would like to use Pycodestyle as a library to get the user's final configuration as a dictionary. Here is what I have tried:

```python
def get_settings(self, args):
    config_file = None
    if args:
        config_file = args[0]

    options, args = process_options(config_file=config_file)
    return options.__dict__
```

This does return the dictionary, but it does not detect the project-local configuration file, only the global one plus whatever explicit configuration file I have given. The documentation is not clear on the meaning of the arguments to `process_options`, so I guess I am missing something there.

Some background: I want to write a [Neovim](https://neovim.io/) plugin which will allow me to read in the final linter configuration so I can set up my editor to use the linter settings for things like maximum text width. Instead of having a separate editor and linter configuration, why not just use the linter configuration directly in the editor? The Neovim side of things is simple enough, it's only the Pycodestyle side I need to figure out.

**Comment by asottile** [comment]
I believe you're looking for [read_config](https://pycodestyle.pycqa.org/en/latest/api.html#pycodestyle.read_config)

**Comment by HiPhish** [comment]
That was my other guess, but I still don't know what parameters to pass.

```python
options, args = process_options(config_file=config_file)
config = read_config(options, args, [], OptionParser())
return config.__dict__
```

**Comment by sigmavirus24** [comment]
pycodestyle isn't intended to be used as a library so we don't document any of the functions here. Can Neovim not parse `ini` files? `pycodestyle` doesn't use anything beyond a very basic parser from python's stdlib. You would merely need to look for the right section in the file and then know how to map pycodestyle setings to neovim ones.

**Comment by sigmavirus24** [comment]
nevermind, I see there are API docs for `read_config`, I don't remember those existing 😆

**Comment by sigmavirus24** [comment]
You probably want to read some internal code to figure out what we're doing https://github.com/PyCQA/pycodestyle/blob/aa3417b6a51f5912e32d9c8c879e1b9dd660d5f8/pycodestyle.py#L2729-L2745 and mimic some of that

**Comment by asottile** [comment]
yeah I was also surprised to find the api docs -- especially because the functions aren't really useful on their own!

**Comment by sigmavirus24** [comment]
Yeah they're actually completely garbage and user-hostile. I wouldn't think an API for pycodestyle would actually be useful. The functions documented would allow someone to write something that "extends" pycodestyle but would need to redistribute and have it's own entry-point rather than properly extending. I'm thinking a pycodestyle major release that dumps this junk is probably a good idea.

**Comment by HiPhish** [comment]
ESLint can be [used as a library](https://eslint.org/docs/developer-guide/nodejs-api), so I figured something similar should be possible with pycodestyle. I could write an INI parser for Neovim, but then I would end up either depending on [lpeg](https://luarocks.org/modules/gvvaughan/lpeg) or writing my own abomination of a parser out of regular expressions with all sorts of edge cases missed. Plus I would have to re-implement the rules for combining options.

Just using pycodestyle would have been really nice. Since you already need Python and pycodestyle installed anyway there would be no extra dependency and all I would need to do is write the glue code that connects pycodestyle to Neovim while pycodestyle does all the actual thinking. And since it uses the user's own pycodestyle the rules would always be in sync with whatever version the user is using.

Would you be willing to reconsider your "not a library" stance?

**Comment by HiPhish** [comment]
Never mind, I'm a moron apparently. Python already comes with the `configparser` module which lets me parse the config files. All I need to do is replicated the rules on in which order to process files. That's definitely more doable and it does not require extra dependencies.

I still think it would be a good idea to have this sort of thing built in, or at least a way to get the list of files to parse, but that's up to you. If you think it's not worth having then feel free to close this issue.

**Comment by HiPhish** [comment]
I just noticed that parsing the config files does not give me the defaults. I have extracted the defaults from `get_parser.__dict__['defaults']` and manually hard-coded it in my plugin. Is there a better way? Anyways, here is what I have come up with:

```python
import os
from configparser import ConfigParser
from pathlib import Path

PROJECT_FILES = ['setup.cfg', 'tox.ini']

# Taken from pycodestyle by evaluating `get_parser.__dict__['defaults']`
DEFAULTS = {
    'verbose': 0,
    'quiet': 0,
    'repeat': True,
    'exclude': '.svn,CVS,.bzr,.hg,.git,__pycache__,.tox',
    'filename': '*.py',
    'select': '',
    'ignore': '',
    # 'show-source': None,
    # 'show-pep8': None,
    # 'statistics': None,
    # 'count': None,
    'max-line-length': 79,
    # 'max-doc-length': None,
    'indent-size': 4,
    # 'hang-closing': None,
    'format': 'default',
    # 'diff': None,
    # 'benchmark': None
}


def parse_file(path: Path, config: ConfigParser) -> ConfigParser:
    if path.exists():
        config.read(path)
    return config


def get_settings(self, args):
    config = ConfigParser()
    config.read_dict({'pycodestyle': DEFAULTS})
    if os.name == 'nt':
        parse_file(Path.home() / '.pycodestyle')
    else:
        conf_dir = os.getenv('XDG_CONFIG_HOME') or (Path.home() / '.config')
        parse_file(Path(conf_dir) / 'pycodestyle', config)
    parse_file(USER_CONF_FILE, config)

    for project_file in PROJECT_FILES:
        path = Path(os.getcwd()) / project_file
        parse_file(path, config)

    return dict(config['pycodestyle'])
```

Not very pretty, but it gets the job done. When looking for the project configuration my code does not search upwards, but that's just a minor issue I am willing to brush under the carpet for now. I am not certain as to what to do about the type of the entries. Currently the parser turns everything into a string, but I could try to parse them into their actual types. Would that make sense? I would have to specify the type or converter function on a per-entry basis.

**Comment by sigmavirus24** [comment]
I mean, I don't think doing this all from scratch is pleasant or going to be sustainable. That said, I also linked you directly to the section of code you'd need to paraphrase to not have to write all of this yourself that would help you understand what to wire together in the currently documented API. That will also take care of types for you if I remember that code correctly. I wouldn't ever want to try to hard code defaults that could change in something since you'll be more likely to be out of sync and wrong at some point (especially given that folks can use different versions of pycodestyle). That said, I'm not really sure what more you expect from us here. Your code isn't wrong. I know nothing about neovim development. I can't tell you what to do.

**Comment by HiPhish** [comment]
Thanks, I have figured out how to get the defaults and handle type conversion. For completeness if anyone comes across this in the future:

```python
# Get a dictionary of default settings
parser: OptionParser = get_parser()
options, __ = parser.parse_args([])
defaults = options.__dict__

# Type conversion is a bit more complicated. We have a `Dict[str, str]` object,
# so we want to parse some of the values into other types. Let's use a key-converter
# dictionary to specify how to convert which entry

def split(s: str) -> List[str]:
    """Split a string on commas. Returns an empty list for the empty string"""
    return s.split(',') if s else []

TYPE_CONVERTERS = {
    'verbose': int,
    'quiet': int,
    'repeat': bool,
    'exclude': split,
    'select': split,
    'ignore': split,
    'show-source': bool,
    'show-pep8': bool,
    'statistics': bool,
    'count': bool,
    'max-line-length': int,
    'max-doc-length': int,
    'indent-size': int,
    'hang-closing': bool,
    'diff': bool,
    'benchmark': bool
}

# Convert the strings to the correct type
for key, converter in TYPE_CONVERTERS.items():
    if key in result and result[key] is not None:
        result[key] = converter(result[key])

```

> I know nothing about neovim development. I can't tell you what to do.

That's OK, I only needed to get the dictionary of setting-value pairs from pycodestyle, from there on I know how to handle the rest and pass it on to Neovim.
